lib/WasmReader/WasmModule.cpp

```diff
@@ -33,8 +33,64 @@ WasmModule::WasmModule(Js::ScriptContext* scriptContext, byte* binaryBuffer, uin
     m_reader->InitializeReader();
 }
 
+uint32
+WasmModule::GetMaxFunctionIndex() const
+{
+    return UInt32Math::Add(GetImportCount(), GetWasmFunctionCount());
+}
+
+Wasm::WasmSignature*
+WasmModule::GetFunctionSignature(uint32 funcIndex) const
+{
+    FunctionIndexTypes::Type funcType = GetFunctionIndexType(funcIndex);
+    if (funcType == FunctionIndexTypes::Invalid)
+    {
+        throw WasmCompilationException(_u("Function index out of range"));
+    }
+
+    funcIndex = NormalizeFunctionIndex(funcIndex);
+    switch (funcType)
+    {
+    case FunctionIndexTypes::Import:
+        return GetSignature(GetFunctionImport(funcIndex)->sigId);
+    case FunctionIndexTypes::Function:
+        return GetWasmFunctionInfo(funcIndex)->GetSignature();
+    default:
+        throw WasmCompilationException(_u("Unknown function index type"));
+    }
+}
+
+FunctionIndexTypes::Type
+WasmModule::GetFunctionIndexType(uint32 funcIndex) const
+{
+    if (funcIndex >= GetMaxFunctionIndex())
+    {
+        return FunctionIndexTypes::Invalid;
+    }
+    if (funcIndex < GetImportCount())
+    {
+        return FunctionIndexTypes::Import;
+    }
+    return FunctionIndexTypes::Function;
+}
+
+uint32
+WasmModule::NormalizeFunctionIndex(uint32 funcIndex) const
+{
+    switch (GetFunctionIndexType(funcIndex))
+    {
+    case FunctionIndexTypes::Import:
+        return funcIndex;
+    case FunctionIndexTypes::Function:
+        Assert(funcIndex >= GetImportCount());
+        return funcIndex - GetImportCount();
+    default:
+        throw WasmCompilationException(_u("Unable to normalize function index"));
```

Comment:
I think I'll go for `Failed function index normalization: function index out of range`